src/decode/lzma.rs

```diff
@@ -264,76 +302,212 @@ where
+    ) -> error::Result<bool> {
+        self.process_next_inner(rangecoder, true)
+    }
+
+    /// Try to process the next iteration of the loop.
+    ///
+    /// This will check to see if there is enough data to consume and advance the
+    /// decompressor. Needed in streaming mode to avoid corrupting the state while
+    /// processing incomplete chunks of data.
+    #[allow(clippy::if_same_then_else)]
+    fn try_process_next<'a>(&mut self, buf: &'a [u8], range: u32, code: u32) -> error::Result<()> {
+        let mut temp = std::io::Cursor::new(buf);
+        let mut rangecoder = rangecoder::RangeDecoder::from_parts(&mut temp, range, code);
+        let _ = self.process_next_inner(&mut rangecoder, false)?;
+        Ok(())
+    }
+
+    pub fn process_mode<'a, R: io::BufRead>(
+        &mut self,
+        mut rangecoder: &mut rangecoder::RangeDecoder<'a, R>,
+        mode: Mode,
+    ) -> error::Result<()> {
+        loop {
+            if let Some(unpacked_size) = self.unpacked_size {
+                if self.output.len() as u64 >= unpacked_size {
+                    break;
+                }
+            } else if match mode {
+                Mode::Run => rangecoder.is_eof()?,
+                Mode::Finish => rangecoder.is_finished_ok()?,
+            } {
+                break;
+            }
+
+            if self.tmp.position() as usize > 0 {
+                // Fill as much of the tmp buffer as possible
+                let start = self.tmp.position() as usize;
+                let bytes_read = rangecoder.read_into(&mut self.tmp.get_mut()[start..])?;
+                let bytes_read = if bytes_read < std::u64::MAX as usize {
+                    bytes_read as u64
+                } else {
+                    return Err(error::Error::LZMAError(
+                        "Failed to convert integer to u64.".to_string(),
+                    ));
+                };
+                self.tmp.set_position(self.tmp.position() + bytes_read);
+                let tmp = *self.tmp.get_ref();
+
+                // Check if we need more data to advance the decompressor
+                if Mode::Run == mode
+                    && (self.tmp.position() as usize) < MAX_REQUIRED_INPUT
+                    && self
+                        .try_process_next(
+                            &tmp[0..self.tmp.position() as usize],
+                            rangecoder.range(),
+                            rangecoder.code(),
+                        )
+                        .is_err()
+                {
+                    return Ok(());
+                }
+
+                // Run the decompressor on the tmp buffer
+                let mut tmp_reader = io::Cursor::new(&tmp[0..self.tmp.position() as usize]);
+                let mut tmp_rangecoder = rangecoder::RangeDecoder::from_parts(
+                    &mut tmp_reader,
+                    rangecoder.range(),
+                    rangecoder.code(),
+                );
```

Comment:
So, rather than exposing all fields and add all these from_parts/to_parts, what is really needed is some kind of `duplicate`/`clone` functionality for the `RangeDecoder` right? There could be a function specifically for `RangeCoder<Cursor<_>>`. I think that would be clearer to have that explicitly in the `RangeDecoder` struct rather than building and destructuring range coders here.

Reply:
Yes, you are cloning the range and code but using a different value for `stream`.

This is a good idea. One issue that I had creating a clone function relates to the type `<R: io::BufRead>`.

Say we have a function called like so:
```rust
let tmp_buf = std::io::Cursor::new("test".as_bytes());
let new_rangecoder = rangecoder.clone_with(tmp_buf);
```

There is a constraint that the new_rangecoder must have the same concrete type `R` as the source of the clone. I.e. `clone_with` only works if both the `rangecoder.stream` and the `tmp_buf` are `R: std::io::Cursor<&[u8]>`. I'll see if anything can be done about this.

Reply:
I resolved the type issue but it is still not possible due to holding multiple references to the rangecoder. Copying the range and code avoids these issues.